packages/design-system/src/scalars/components/amount-field/use-amount-field.ts

```diff
@@ -1,40 +1,137 @@
-import { AmountCurrency, AmountFieldPropsGeneric, AmountValue } from "./types";
-import { getLabelValueCurrenct } from "./utils";
+import {
+  AmountCurrency,
+  AmountFieldPropsGeneric,
+  AmountToken,
+  AmountValue,
+} from "./types";
+import { getCountryCurrencies, getTokens } from "./utils";
 
 interface UseAmountFieldProps {
   value?: AmountValue;
   defaultValue?: AmountValue;
   type: AmountFieldPropsGeneric["type"];
   allowedCurrencies?: string[];
+  allowedTokens?: string[];
+  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
+  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => void;
 }
 
 export const useAmountField = ({
   value,
   defaultValue = 0,
   type,
   allowedCurrencies = [],
+  allowedTokens = [],
+  onChange,
+  onBlur,
 }: UseAmountFieldProps) => {
   const currentValue = value ?? defaultValue;
 
   const valueInput =
     type === "Amount" || type === "AmountPercentage"
       ? (currentValue as number)
-      : (currentValue as AmountCurrency).amount;
+      : (currentValue as AmountCurrency | AmountToken).amount;
+
   const isPercent = type === "AmountPercentage";
-  const isCurrency = type === "AmountCurrency";
 
-  const options = getLabelValueCurrenct(allowedCurrencies);
-  const currency =
-    type === "AmountCurrency" ? (currentValue as AmountCurrency).currency : "";
+  //Allow select only if type is AmountCurrency or AmountToken
+  const isShowSelect = type === "AmountCurrency" || type === "AmountToken";
+
+  // Filter only if allowedCurrencies is provided
+  const optionsCurrencies = getCountryCurrencies(allowedCurrencies);
+  const optionsTokens = getTokens(allowedTokens);
+
+  const options = type === "AmountCurrency" ? optionsCurrencies : optionsTokens;
+
+  const valueSelect =
+    type === "AmountCurrency"
+      ? (currentValue as AmountCurrency).currency
+      : (currentValue as AmountToken).token;
+
+  // Handle the change of the input
+  const handleOnChangeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
+    let newValue: AmountValue = {} as unknown as AmountValue;
+    if (type === "AmountCurrency" && typeof value === "object") {
+      newValue = {
+        ...value,
+        amount: Number(e.target.value),
+      } as AmountValue;
+    }
+    if (type === "AmountToken" && typeof value === "object") {
+      newValue = {
+        ...value,
+        amount: BigInt(e.target.value),
+      } as AmountValue;
+    }
+    if (type === "Amount" || type === "AmountPercentage") {
+      newValue = Number(e.target.value);
+    }
+    //Create the event
+    const nativeEvent = new Event("change", {
+      bubbles: true,
+      cancelable: true,
+    });
+    Object.defineProperty(nativeEvent, "target", {
+      value: { value: newValue },
+      writable: false,
+    });
+    onChange?.(newValue as unknown as React.ChangeEvent<HTMLInputElement>);
```

Comment:
that case doesn't make sense there as we're passing the value instead of the event

Reply:
You're right—thank you for catching that. Done